docs/architecture.md

```diff
@@ -0,0 +1,473 @@
+```
+
+### Block Producer
+
+#### Ports: fuel_block_producer::ports
+
+```rust
+trait TransactionPool {
+    fn select_block_txs(max_gas: u64, block_height: BlockHeight) -> Result<Vec<Arc<CheckedTransaction>>>;
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Executor {
+    async fn produce_block(block: &mut FuelBlock) -> Result<()>;
+}
+
+trait Relayer {
+    fn get_finalized_da_height() -> Result<DaHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::block_producer
+```rust
+impl fuel_poa_consensus::ports::BlockProducer for Service<BlockProducer> {
+    // insert impl here
+}
+```
+
+### Block Importer
+
+#### Ports: fuel_block_importer::ports
+
+```rust
+trait Relayer {
+    // wait until the relayer is synced with ethereum
+    async fn await_synced() -> Result<()>;
+    // wait until a specific da_height is reached
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+}
+
+trait Executor {
+    async fn validate_and_store_block(block: &FuelBlock) -> Result<()>;
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+
+```rust
+trait PeerToPeer {
+    type SealedHeaderResponse: NetworkData<SealedFuelBlockHeader>;
+    type BlockResponse: NetworkData<Vec<SealedFuelBlock>>;
+    type GossipedBlockHeader: NetworkData<SealedFuelBlockHeader>;
+    
+    async fn fetch_best_network_block_header() -> Result<Self::SealedHeaderResponse>;
+    async fn fetch_blocks(query: Range<BlockHeight>) -> Result<Self::BlockResponse>;
+    // punish the sender for providing an invalid block header
+    fn report_invalid_block_header(invalid_header: &Self::SealedHeaderResponse) -> Result<()>;
+    // punish the sender for providing a set of blocks that aren't valid
+    fn report_invalid_blocks(invalid_blocks: &Self::BlockResponse) -> Result<()>;
+    // await a newly produced block from the network (similar to stream.next())
+    async fn next_gossiped_block_header() -> Result<Self::SealedFuelBlockHeader>;
+    // notify the p2p network whether to continue gossiping this message to others or
+    // punish the peer that sent it
+    fn notify_gossip_block_validity(message: &Self::GossipedBlockHeader, validity: GossipValidity);
+}
+
+// Generic wrapper for data received from the network
+pub trait NetworkData<D> {
+    // Transfer ownership of the provided data without consuming other message metadata
+    fn take_data(&mut self) -> Option<D>;
```

Comment:
What does this option represent? Is this saying that all network data is optional. If so that answers my above question.

Reply:
The intention of this type is to hide p2p metadata like network message ids or peer ids needed for reporting back message validity. I could get rid of the option and just have an `AsRef<D>` to access the message payload, but that would force consumers to always have to clone the data.

Reply:
I think I'm still missing something. If it can be `AsRef<D>` why can't it just be `-> D`?

Reply:
It was just a way to be able to move the data out of this wrapper without cloning, but maybe there are better ways.

Ultimately this isn't intended to be the exact design as added in the disclaimer above. I think it will be more clear during implementation.